core/src/main/scala-3/scairdl/IRElements.scala

```diff
@@ -178,29 +180,83 @@ case class Assemblyformat(
 
 // Seq[FormatDirective]
 
-// trait FormatDirective {
-//   def parse(parser: Parser): P[Any]
-// }
+trait FormatDirective {
+  def parse[$: P]: P[Any]
+}
 
-// case class LiteralDirective(literal: String) extends FormatDirective {
-//   override def parse(parser: Parser): P[Unit] = P(literal)
-// }
+object NewParser {
 
-// case class OperandDirective(name: String) extends FormatDirective {
-//   override def parse(parser: Parser): P[String] = P(Parser.ValueUse)
-// }
+  def resultTypeDirective[$: P]: P[ResultTypeDirective] =
+    P("type($result)").map(_ => ResultTypeDirective())
 
-// case class TypeDirective(inner: OperandDirective) extends FormatDirective {
-//   override def parse(parser: Parser): P[String] = P(parser.Type)
-// }
+  def operandDirective[$: P]: P[OperandDirective] =
+    P("$" ~ CharsWhileIn("a-zA-Z0-9_").!)
+      .filter(_ != "result")
+      .map(OperandDirective)
 
-// //not sure
-// case class AssemblyFormat(directives: Seq[FormatDirective]) {
+  def typeDirective[$: P]: P[TypeDirective] =
+    P("type(" ~ (operandDirective) ~ ")").map(TypeDirective)
 
-// def parse(parser: Parser): P[Seq[Any]] =
-//     P(directives.map(_.parse(parser)).reduceOption(_ ~ _).getOrElse(("")))
+  def literalDirective[$: P]: P[LiteralDirective] =
+    P(CharIn("`,:`").!).map(LiteralDirective)
+
+  def formatDirective[$: P]: P[FormatDirective] =
+    P(typeDirective | operandDirective | literalDirective | resultTypeDirective)
+
+  def assemblyFormat[$: P]: P[AssemblyFormat] =
+    P(formatDirective.rep(1)).map(directives => AssemblyFormat(directives))
+
+  def parseFormat(input: String): Parsed[AssemblyFormat] =
+    parse(input, assemblyFormat(_))
+
+}
+
+case class LiteralDirective(literal: String) extends FormatDirective {
+  override def parse[$: P]: P[Unit] = P(literal)
+}
+
+case class ResultTypeDirective(name: String = "result")
+    extends FormatDirective {
+
+  override def parse[$: P]: P[String] =
+    P("type($result)").map(_ => "result_type")
+
+}
+
+case class OperandDirective(name: String) extends FormatDirective {
+
+  override def parse[$: P]: P[String] =
+    P("$" ~ CharsWhileIn("a-zA-Z0-9_").!).filter(_ != "result")
+
+}
+
+object OperandDirective {
+
+  def parse[$: P]: P[OperandDirective] =
+    P("$" ~ CharsWhileIn("a-zA-Z0-9_").!)
+      .filter(_ != "result")
+      .map(OperandDirective(_))
```

Comment:
I think all those are redundant now, right? As you parse them in the NewParser itself anyway.
I would remove the `def parse` in the FormatDirective trait and all those implementations then, unless I missed something?

Reply:
Yes I think it would be nice!